Add hack/verify-all.sh and make verify

pmorie opened the pull request:
Depends on #96 and #97

martinmaly commented on Makefile:
```diff
@@ -48,3 +48,17 @@ format:
+#   make verify BRANCH=branch_x
+.PHONY: verify
+verify:
+	KUBE_VERIFY_GIT_BRANCH=$(BRANCH) hack/verify-all.sh -v
```
the KUBE_VERIFY_GIT_BRANCH seems unused. what is it for?

pmorie replied:
Lifted from upstream -- I believe we can remove it for now.  I'll take it out.

duglin replied:
Did you want to remove this?

martinmaly commented on contrib/registry/data/types/test/test-instance.py:
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
```
Github won't let me comment on the *.pyc files ... an we remove them from the repo, please?

martinmaly commented on hack/utilities.sh:
```diff
@@ -75,3 +75,12 @@ function retry() {
+
+# Some useful colors.
+if [[ -z "${color_start-}" ]]; then
+  declare -r color_start="\033["
```
I realize the [Google shell style guide](https://google.github.io/styleguide/shell.xml?showone=Constants_and_Environment_Variable_Names#Constants_and_Environment_Variable_Names) may not have much bearing here although I did find it quite useful, even insightful in my quest to learn more about bash. for constants it recommends ALL_CAPS names. See what you think...

martinmaly replied:
Yes. I think in order to defer to the style guide as an authoritative guide that governs our scripts, we'll need to agree to it as a community.

My intention was not to defer to it (i.e. use the guide as a justification to demand a change). I merely pointed out its existence, observed I'd found it useful and even highlighted that it may not have any bearing (anticipating need for agreement before it can be used in any meaningful way).

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+
+#!/bin/bash
+
+# Copyright 2014 The Kubernetes Authors.
```
dual header (this one with 2014 year)

pmorie replied:
oops.

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#!/bin/bash
```
rm

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+run-checks "${ROOT}/hack/verify-*.sh" bash
+exit ${ret}
+
+# ex: ts=2 sw=2 et filetype=sh
```
just curious ... what is this file suffix for?

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+  esac
+done
+
+if ${SILENT} ; then
```
I worry this may misbehave if SILENT is the default value because ${SILENT} will expand to empty string, yielding:

if ; then
...
fi

which is invalid bash syntax ?
(I am not bash expert so there is a chance something else will happen and this is OK).

pmorie replied:
This is lifted from upstream, but should be changed.  I would rather wait until we have this stuff extracted into a separate repo and submodule it in than changing it locally and having this get out of sync w/ upstream.  Does that work for you?

martinmaly replied:
sounds good. One can wish the upstream repo had higher quality but oh well. I will look forward to the repo refactoring where we can bring all scripts in as a simple dependency

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+  done
+}
+
+while getopts ":v" opt; do
```
(no need to change, only sharing where my bash coding has evolved to). Ultimately I decided that getopts with its short names is a bit suboptimal so I started parsing arguments [by hand](https://github.com/kubernetes-incubator/service-catalog/blob/master/hack/build.sh#L25)

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+      echo -e "${color_green}SUCCESS${color_norm}  ${t}\t${elapsed}s"
+    else
+      echo -e "${color_red}FAILED${color_norm}   ${t}\t${elapsed}s"
+      ret=1
```
is this a special variable? specifically, is it going to be returned from the function?

martinmaly commented on hack/verify-all.sh:
```diff
@@ -0,0 +1,106 @@
+
+ret=0
+run-checks "${ROOT}/hack/verify-*.sh" bash
+exit ${ret}
```
ah, I see. the ret in the function is global. This is fine (by me) though if I were coding it up I'd have the function return 0 or 1 and here do:

run_checks ... \
  || error_exit "stuff failed"

(disclaimer - my bash skills are mostly self-taught so I am likely not an idiomatic bash programmer)

krancour commented:
Can I ask that this PR also update `.travis.yml` to do this:

```
make verify build test lint coverage
```

This will guarantee verification runs during CI since we can't reliably count on developers to have installed this as a commit hook. It will also short circuit the remainder of the build and test (to conserve resources) when verification has failed.

pmorie commented:
@krancour yes, I will update the travis config.

pmorie commented:
@krancour rebased and added to travis build.  Let's see if our travis actually likes it.

duglin commented on Makefile:
```diff
@@ -32,6 +32,7 @@ SUB := $(addsuffix .sub, $(ALL))
 build: build.sub
 build-linux: build-linux.sub
 build-darwin: build-darwin.sub
+clean: clean.sub
```
I think this is a dup of line 41

pmorie replied:
Ack, will address

duglin commented:
I think there are some questions from @martinmaly that are unanswered. 
I have just one requested change.

pmorie commented:
@martinmaly

Regarding all as-yet unaddressed comments: this script is lifted from k8s and should be replaced by submodule containing the k8s-generic version once it is split out into its own repo.  So, that being the case, I would prefer to make changes to the k8s-generic version once it is extracted.

With that said, some thoughts:

1.  There is a complete lack of consistency around bash style in k8s; it seems to depend entirely on the reviewer and has been a major source of frustration to me in the past
2.  I am all about following best practices, but I must admit there seems to be some significantly disagreement or variance as to what good bash practice is (up to and including, don't use bash)
3.  A style guide would be very very helpful, IMO

Thoughts?

arschles commented on pkg/apiserver/apiserver.go:
```diff
@@ -1,3 +1,19 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
I'd prefer this and the below comment get submitted in a separate PR, but since I'd rather us get this PR merged I'm not gonna hold this up.

arschles commented:
This is fine. LGTM2

duglin commented:
rebase needed

martinmaly commented on Makefile:
```diff
@@ -81,3 +81,18 @@ coverage:
 	go install -v github.com/kubernetes-incubator/service-catalog/cmd/service-catalog
 	go build -v -o apiserver cmd/service-catalog/server.go
+
+verify:
```
looks like we don't need line 85?

pmorie commented:
I'm going to close this out and re-open a PR that uses repo-infra.
